src/main/java/com/evpa/DeadlockExample.java

```diff
@@ -0,0 +1,39 @@
+package com.evpa;
+
+public class DeadlockExample {
+
+    private Object process = new Object();
+    private boolean stopped = true;
+
+    public void run() {
+        synchronized (process) {
+            isStopped();
+        }
+    }
+
+    private synchronized boolean isStopped() {
+        return stopped;
+    }
+
+    public synchronized void stopProcess() {
+        synchronized (process) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 The code acquires two synchronization objects in **two different orders**, which can cause **circular wait** (deadlock).

 The **first order** is: first acquire `DeadlockExample.this` and then `process`:

- enter `stopProcess()` at** line 18**

- **acquire synchronization** on `DeadlockExample.this` at entry to synchronized method `stopProcess()`, see** line 18**

- **acquire synchronization** on `process` at** line 19**



 The **second order** is: first acquire `process` and then `DeadlockExample.this`:

- enter `run()` at** line 8**

- **acquire synchronization** on `process` at** line 9**

- call `isStopped()` at** line 10**

- **acquire synchronization** on `DeadlockExample.this` at entry to synchronized method `isStopped()`, see** line 14**



**Fix**  
 The fix is to acquire synchronization in only one order. Typically you can do this by adding extra synchronization to the synchronization order you want to prohibit.